src/rpc/mining.cpp

```diff
@@ -118,29 +148,54 @@ static UniValue generateBlocks(const CTxMemPool& mempool, const CScript& coinbas
         if (!pblocktemplate.get())
             throw JSONRPCError(RPC_INTERNAL_ERROR, "Couldn't create new block");
         CBlock *pblock = &pblocktemplate->block;
-        {
-            LOCK(cs_main);
-            IncrementExtraNonce(pblock, ::ChainActive().Tip(), nExtraNonce);
-        }
-        while (nMaxTries > 0 && pblock->nNonce < std::numeric_limits<uint32_t>::max() && !CheckProofOfWork(pblock->GetHash(), pblock->nBits, Params().GetConsensus()) && !ShutdownRequested()) {
-            ++pblock->nNonce;
-            --nMaxTries;
-        }
-        if (nMaxTries == 0 || ShutdownRequested()) {
+
+        uint256 block_hash;
+        if (!GenerateBlock(*pblock, nMaxTries, nExtraNonce, block_hash)) {
             break;
         }
-        if (pblock->nNonce == std::numeric_limits<uint32_t>::max()) {
-            continue;
```

Comment:
This change means that the extra_nonce logic is now unused. Previously if the nNonce field rolled, we'd go around the loop again and call `IncrementExtraNonce()`. Now we'll just return `true`.

I've proposed removing that logic entirely in #21533.

Reply:
Never mind, I misunderstood the new interface for `GenerateBlock()`:

- returns false: generation of block failed
- returns true with valid block_hash out-param: generation of block succeeded
- returns true with invalid (nullified) block_hash out-param: generation of block didn't succeed, maximum nNonce reached

(thanks to @theStack for pointing this out: https://github.com/bitcoin/bitcoin/pull/21533#pullrequestreview-634095912)